volume/driver/overlay_linux.go

```diff
@@ -78,21 +92,99 @@ func (driver *OverlayDriver) CreateCopyOnWriteLayer(path string, parent string)
 		workDir,
 	)
 
+	fmt.Println(path, opts)
 	return syscall.Mount("overlay", path, "overlay", 0, opts)
 }
 
-func (driver *OverlayDriver) layerDir(path string) string {
-	return filepath.Join(driver.OverlaysDir, driver.pathId(path))
+// We apply 2 types of mounts when using the overlay driver
+//
+// 1. For any new volume, we create its root in the overlaysDir and
+// 	  issue a bind mount of that root into the liveVolumesDir
+// 2. For COW volumes, we create upper and work dirs and use
+// 	  ancestry to determine the lower dir. These dirs are
+//	  used to create an overlay mount.
+//
+// These mounts can disappear when the system reboots (mount table cleared)
+// As a precaution we reattach mounts during startup to fix missing ones
+func (driver *OverlayDriver) RecoverMountTable(liveVolumesDir string) error {
+	liveVolumes, err := ioutil.ReadDir(liveVolumesDir)
+	if err != nil {
+		return err
+	}
+
+	for _, volumeFileInfo := range(liveVolumes) {
+
+		liveVolumePath := filepath.Join(liveVolumesDir, volumeFileInfo.Name())
+		liveVolumeDataPath := filepath.Join(liveVolumePath, "volume")
+		parentSymlink := filepath.Join(liveVolumePath, "parent")
+
+		// a parent symlink indicates a COW
+		if _, err := os.Stat(parentSymlink); !os.IsNotExist(err) {
```

Comment:
I'm a bit confused with this one 🤔 

Does this mean that "for any `err` that is not `ENOENT`, go ahead with the following"? I.e., "go ahead for EACCES, ENAMETOOLONG, ENOMEM, etc etc"

Reply:
Yeah it does, I placed it here as a safety check to see that the parent symlink exists. The symlink is always named "parent" and is created by the baggage claim process. So I think we're cool about permission denied or name too long errors. If you run out of memory you probably have bigger problems than this. I think it's reasonable to assume that IsExists is the only error to care for?

Reply:
fixed lemme know what you think